### DISCUSSION TITLE
PNGs: Too many text-chunks

### DISCUSSION THREAD

**Discussion by joshuamsager** [libvips/libvips#2418] [Help wanted]
8.11.3 introduced a bunch of failures in our application for [PNGs with text chunks greater than 10](https://github.com/libvips/libvips/commit/526368651175bb8a916bcfb794599e12f5e57237). 

Sample image:
```
joshuasager@~  $ identify -verbose ~/Downloads/test.png | grep png
  Filename: /Users/joshuasager/Downloads/test.png
  Mime type: image/png
    png:bKGD: chunk was found (see Background color, above)
    png:cHRM: chunk was found (see Chromaticity, above)
    png:iCCP: chunk was found
    png:IHDR.bit-depth-orig: 8
    png:IHDR.bit_depth: 8
    png:IHDR.color-type-orig: 2
    png:IHDR.color_type: 2 (Truecolor)
    png:IHDR.interlace_method: 0 (Not interlaced)
    png:IHDR.width,height: 80, 80
    png:pHYs: x_res=11811, y_res=11811, units=1
    png:text: 28 tEXt/zTXt/iTXt chunks were found
    png:text-encoded profiles: 4 were found
    png:tIME: 2020-10-06T18:14:30Z
```

I understand that this was implemented in order to close an attack vector... but restricting PNGs to 10 chunks is a big limitation for our use case. What is the recommended way to resolve this issue?

**Comment by jcupitt** [comment]
Hi @joshuamsager,

We should clearly raise this limit.

Do you have any feeling for what a useful upper limit might be?

**Comment by joshuamsager** [comment]
> Do you have any feeling for what a useful upper limit might be?

The challenge there is that a limit that fits my use-case may not fit other developers' use cases. I'm wondering if there is a mechanism to allow the developer to compile libvips with a specific flag to change this limit. I'm ok with keeping 10 as a default. 

Does this pattern already exist within libvips? (I'm new to the repo and haven't quite learned my way around just yet)

**Reply by jcupitt** [reply]
We have an `unlimited` flag to some loaders to disable DoS attack limits, perhaps the PNG loader should use this too.

Would you mind explaining why you have so many? I'm not sure I've seen a real PNG image with more than 1 or 2 text chunks.

**Reply by joshuamsager** [reply]
The service my team runs allows our customers to upload PNGs which we then serve (after applying various transformations via libvips). From what I have seen, when our customers create PNGs using Adobe Photoshop, it creates an image with many text chunks. I do not know what in particular causes so many chunks to be created, however.

Ill take a closer look at the `unlimited` flag.

**Comment by joshuamsager** [comment]
Rudimentary PR here: https://github.com/libvips/libvips/pull/2419/
